### DISCUSSION TITLE
Evaluate Profile-Guided Optimization (PGO) and Post Link Optimization (PLO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [dathere/qsv#1433] [Ideas]
Hi!

Recently I checked Profile-Guided Optimization (PGO) improvements on multiple projects. The results are available [here](https://github.com/zamazan4ik/awesome-pgo/). According to the tests, PGO can help with achieving better performance. Also, I found interesting [results](https://github.com/eBay/tsv-utils/blob/master/docs/lto-pgo-study.md) about PGO effects on `tsv-utils` - project in the similar domain as `qsv`. Since all of these, I think trying to optimize qsv with PGO can be a good idea.

I already did some benchmarks and want to share my results.

## Test environment

* Macbook Pro 14 M1 (6 + 2 CPU, 16 Gib RAM)
* Compiler - Rustc 1.74
* qsv version: `master` branch on commit `531acbb072c48cbaca5d58b593243e0f5f0ec8d3`

Right now I cannot perform the tests on my Linux machine (Fedora-based) due to some build errors: https://github.com/jqnatividad/qsv/issues/1431 . But I think the results should be the same for the Linux platform as well.

## Benchmark

For benchmark purposes, I use [this](https://github.com/jqnatividad/qsv/blob/master/scripts/benchmarks.sh) QSV benchmark. For PGO optimization I use [cargo-pgo](https://github.com/Kobzol/cargo-pgo) tool. The same benchmark suite was used for the PGO training phase built with `cargo pgo build -- --release --locked -F feature_capable,apply,geocode,luau,to,polars --bin qsv` but with disabled LTO. The only change to the benchmark suite was done is benchmark run reduction since for the training phase is enough to run every test case only once.

PGO optimized results I got with QSV built with `cargo pgo optimize build -- --release --locked -F feature_capable,apply,geocode,luau,to,polars --bin qsv` but with disabled LTO. Release version is built with `cargo build --release --locked -F feature_capable,apply,geocode,luau,to,polars --bin qsv`.

Unfortunately, due to the [bug](https://github.com/rust-lang/rust/issues/115344#issuecomment-1819911153) in the Rustc compiler right now PGO cannot be enabled simultaneously with LTO for QSV. So I compare "QSV with LTO" vs "QSV with PGO". Later, when the bug will be fixed, we can apply LTO + PGO to QSV at the same time.

## Results

I got the following results:

* QSV Release + LTO results: https://gist.github.com/zamazan4ik/f166967f12ad27a2bf4253975c2e1907
* QSV Release + PGO optimized results: https://gist.github.com/zamazan4ik/25611016ee295e13b29da4c07adb681b
* (just for reference) QSV Relese + PGO instrumentation: https://gist.github.com/zamazan4ik/a0d217d3e25ff23670eae20003ddd40c

As I interpret the results, PGO measurably improves QSV performance in many cases.

## Further steps

I can suggest the following action points:

* Perform more PGO benchmarks on QSV. If it shows improvements - add a note to the documentation about possible improvements in QSV performance with PGO.
* Providing an easier way (e.g. a build option) to build scripts with PGO can be helpful for the end-users and maintainers since they will be able to optimize QSV according to their workloads.
* Optimize pre-built QSV binaries

Testing Post-Link Optimization techniques (like [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md)) would be interesting too (Clang and Rustc already use BOLT as an addition to PGO) but I recommend starting from the usual PGO.

Here are some examples of how PGO optimization is integrated in other projects:

* Rustc: a CI [script](https://github.com/rust-lang/rust/blob/master/src/ci/stage-build.py) for the multi-stage build
* GCC:
  - Official [docs](https://gcc.gnu.org/install/build.html), section "Building with profile feedback" (even AutoFDO build is supported)
  - A [part](https://github.com/gcc-mirror/gcc/blob/4832767db7897be6fb5cbc44f079482c90cb95a6/configure#L7818) in a "wonderful" `configure` script 
* Clang: [Docs](https://llvm.org/docs/HowToBuildWithPGO.html) 
* Python: 
  - CPython: [README](https://github.com/python/cpython#profile-guided-optimization)
  - Pyston: [README](https://github.com/pyston/pyston#building)
* Go: [Bash script](https://github.com/golang/go/blob/master/src/cmd/compile/profile.sh)
* V8: [Bazel flag](https://github.com/v8/v8/blob/main/BUILD.gn#L184)
* ChakraCore: [Scripts](https://github.com/chakra-core/ChakraCore/tree/master/Build/scripts/pgo)
* Chromium: [Script](https://chromium.googlesource.com/chromium/src/build/config/+/refs/heads/main/compiler/pgo/BUILD.gn)
* Firefox: [Docs](https://firefox-source-docs.mozilla.org/build/buildsystem/pgo.html)
   - Thunderbird has PGO support too
* PHP - [Makefile command](https://github.com/php/php-src/blob/master/build/Makefile.global#L138) and old Centminmod [scripts](https://github.com/centminmod/php_pgo_training_scripts)
* MySQL: [CMake script](https://github.com/mysql/mysql-server/blob/8.0/cmake/fprofile.cmake)
* YugabyteDB: [GitHub commit](https://github.com/yugabyte/yugabyte-db/commit/34cb791ed9d3d5f8ae9a9b9e9181a46485e1981d)
* FoundationDB: [Script](https://github.com/apple/foundationdb/blob/1a6114a66f3de508c0cf0a45f72f3687ba05750c/contrib/generate_profile.sh)
* Zstd: [Makefile](https://github.com/facebook/zstd/blob/dev/programs/Makefile#L232)
* [Foot](https://codeberg.org/dnkl/foot): [Scripts](https://codeberg.org/dnkl/foot/src/branch/master/pgo)
* Windows Terminal: [GitHub PR](https://github.com/microsoft/terminal/pull/10071)
* Pydantic-core: [GitHub PR](https://github.com/pydantic/pydantic-core/pull/741)
* file.d: [GitHub PR](https://github.com/ozontech/file.d/pull/469)
* OceanBase: [CMake flag](https://github.com/oceanbase/oceanbase/blob/master/cmake/Env.cmake#L55)

**Comment by jqnatividad** [comment]
Amazing stuff @zamazan4ik !

I've been micro-optimizing the code and squeezing out more performance with every release but have been putting off PGO as I thought that the prep work would be prohibitive.

Using [samply](https://github.com/mstange/samply/) was my next major step in my micro-optimization efforts. That's why I created a samply release profile.

Thankfully, you made the connection that qsv's benchmark suite would be the perfect way to create a PGO pipeline - not only for the prebuilt binaries, but also for folks who want to compile and tune their own PGO custom builds based on their data, workload and target platforms!

And the payoff is eye-popping! Some commands are 20-30% faster!

I'll do as per your suggested action points.  This will be a great way to put a cherry on top of qsv as I eye a 1.0 release in the next few months...

Thanks heaps!!!
